driver-core/src/main/com/mongodb/connection/netty/ReadTimeoutHandler.java

```diff
@@ -40,13 +40,16 @@ public ReadTimeoutHandler(final long readTimeout) {
     }
 
     void scheduleTimeout(final ChannelHandlerContext ctx) {
-        timeout = ctx.executor().schedule(new ReadTimeoutTask(ctx), readTimeout, TimeUnit.MILLISECONDS);
+        if (timeout == null) {
+            timeout = ctx.executor().schedule(new ReadTimeoutTask(ctx), readTimeout, TimeUnit.MILLISECONDS);
+        }
     }
 
     void removeTimeout(final ChannelHandlerContext ctx) {
         if (ctx.channel().eventLoop().inEventLoop()) {
             if (timeout != null) {
```

Comment:
Even for the inEventLoop check?  I think that this flow is possible:

* readAsync called
* scheduleReadTimeout called, timeout scheduled
* bytes are already available, so no pendingReader set
* removeTimeout called, but we're not in the event loop, so timeout is not cancelled